<a href="https://colab.research.google.com/github/sergiocostaifes/PPCOMP_DM/blob/main/notebooks/01_ingest_validate.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# NB01 — Ingestão e Validação dos Dados

## 1. Contexto

Este notebook constitui a primeira etapa operacional do pipeline PPCOMP_DM após a análise exploratória inicial do dataset bruto realizada no NB00.

Enquanto o NB00 caracteriza a fonte primária BORG/Kaggle de forma exploratória, o NB01 realiza a ingestão controlada do arquivo `borg_traces_data.csv`, valida sua estrutura básica, trata inconsistências temporais evidentes e persiste uma versão validada e otimizada dos dados para as etapas seguintes.

No contexto desta pesquisa, os dados representam eventos operacionais de uma infraestrutura computacional de larga escala. Esses eventos serão posteriormente normalizados temporalmente, agregados em janelas de 5 minutos, utilizados para detecção de criticidade, transformados em atributos preditivos e empregados na formulação supervisionada de antecipação de falhas.

## 2. Objetivo

O objetivo principal deste notebook é produzir uma base validada, tipada e rastreável a partir do CSV bruto.

Especificamente, este notebook busca:

1. Montar o ambiente de execução no Google Colab;
2. Garantir o acesso ao repositório do projeto;
3. Carregar os caminhos padronizados definidos em `src.paths`;
4. Ler o arquivo bruto `borg_traces_data.csv`;
5. Remover a coluna auxiliar `Unnamed: 0`, quando presente;
6. Validar a existência da coluna temporal `time`;
7. Converter `time` para formato numérico;
8. Identificar e tratar valores temporais extremos usados como sentinela;
9. Verificar estatísticas temporais básicas, como mínimo, mediana e máximo;
10. Verificar a consistência semântica entre `event == "FAIL"` e `failed == 1`;
11. Selecionar as colunas necessárias para as etapas posteriores;
12. Ajustar tipos de dados para reduzir uso de memória;
13. Remover registros sem `time` válido;
14. Persistir o arquivo validado `trace_raw_validated.parquet`;
15. Gerar o resumo `01_ingest_validate_summary.json`.

## 3. Papel no Pipeline

O NB01 é uma etapa de ingestão e validação estrutural.

Ele não realiza agregação temporal, não calcula taxa de falha por janela, não detecta episódios críticos, não cria atributos temporais e não realiza modelagem.

Seu papel é garantir que o dataset bruto possa ser lido de forma reprodutível e convertido em uma base mínima confiável para os notebooks posteriores.

Em particular, o NB01 fornece a entrada direta para:

| Notebook | Papel posterior |
|---|---|
| NB02 | Limpeza e normalização temporal |
| NB03 | Construção da série temporal em janelas de 5 minutos |
| NB04 | Cálculo de criticidade e detecção de episódios |
| NB05 | Engenharia de atributos |
| NB06 | Rotulagem de estados e formulação supervisionada |

Assim, o NB01 deve ser interpretado como a fundação operacional do pipeline analítico.

## 4. Artefatos Utilizados

Este notebook utiliza como entrada primária:

| Artefato | Origem | Finalidade |
|---|---|---|
| `borg_traces_data.csv` | Dados brutos BORG/Kaggle | Fonte primária da ingestão |

O notebook gera os seguintes artefatos:

| Artefato | Diretório | Finalidade |
|---|---|---|
| `trace_raw_validated.parquet` | `PROCESSED_PATH` | Dataset validado e tipado para o pipeline |
| `01_ingest_validate_summary.json` | `REPORTS_PATH` | Resumo rastreável da ingestão e validação |

## 5. Estrutura do Notebook

O notebook está organizado em uma célula consolidada de código, dividida internamente em blocos numerados.

A célula realiza:

1. Bootstrap do ambiente Colab;
2. Configuração do repositório;
3. Importação dos caminhos do projeto;
4. Instalação/verificação de dependências;
5. Leitura diagnóstica do CSV bruto;
6. Remoção da coluna auxiliar `Unnamed: 0`;
7. Tratamento de valores sentinela na coluna `time`;
8. Verificação de estatísticas temporais básicas;
9. Verificação de concordância entre `event` e `failed`;
10. Seleção das colunas mantidas no Parquet validado;
11. Ajuste de tipos de dados;
12. Remoção de registros sem tempo válido;
13. Persistência do Parquet validado;
14. Geração do summary JSON;
15. Persistência do resumo da etapa em `01_ingest_validate_summary.json`.

## 6. Observações Metodológicas

Este notebook não remove os registros com `time == 0`.

A presença de `time == 0` é apenas registrada nesta etapa. A decisão metodológica sobre preservação, interpretação e eventual análise de sensibilidade do bloco inicial é tratada no NB02 e nos notebooks posteriores.

Os registros removidos nesta etapa são apenas aqueles sem valor temporal válido após o tratamento de sentinela. Na execução atual, isso corresponde a 3 registros.

A conversão de tipos tem finalidade operacional e de eficiência de armazenamento. Ela não altera a semântica dos eventos nem redefine variáveis analíticas.

As colunas `average_usage`, `maximum_usage` e `random_sample_usage` são mantidas na seleção estendida, mas sua utilidade posterior depende da capacidade de conversão e aproveitamento efetivo nos notebooks seguintes.

## 7. Critério de Validação

A etapa será considerada válida se:

1. O arquivo bruto for encontrado e lido corretamente;
2. A coluna `time` estiver presente;
3. Valores temporais sentinela forem identificados e removidos por meio de `dropna`;
4. A concordância entre `event == "FAIL"` e `failed == 1` for verificada;
5. O arquivo `trace_raw_validated.parquet` for gerado;
6. O arquivo `01_ingest_validate_summary.json` for gerado;
7. O número de registros persistidos for compatível com a remoção dos registros temporais inválidos.

Na execução de referência, o CSV bruto possui 405.894 registros após a remoção da coluna auxiliar, e o Parquet validado preserva 405.891 registros.

In [1]:
# ============================================================
# NB01 — Ingestão e Validação dos Dados
# Pipeline PPCOMP_DM
#
# CÉLULA ÚNICA CONSOLIDADA — versão coerente com abril/2026
#
# Escopo:
# - Lê diretamente o CSV bruto borg_traces_data.csv
# - Remove a coluna auxiliar Unnamed: 0, quando presente
# - Trata valores temporais sentinela na coluna time
# - Verifica consistência entre event == "FAIL" e failed == 1
# - Seleciona as colunas mantidas no pipeline
# - Ajusta tipos de dados conforme a versão histórica de abril
# - Remove registros sem time válido
# - Persiste trace_raw_validated.parquet
# - Persiste 01_ingest_validate_summary.json
#
# Importante:
# - Este notebook não realiza agregação temporal
# - Este notebook não calcula criticidade
# - Este notebook não detecta episódios
# - Este notebook não cria features temporais
# - Este notebook não realiza modelagem
# - Registros com time == 0 são preservados para tratamento/interpretação no NB02
# ============================================================

# ─────────────────────────────────────────────────────────────
# BLOCO 1 — Bootstrap padrão PPCOMP_DM no Colab
# ─────────────────────────────────────────────────────────────

from google.colab import drive
drive.mount("/content/drive", force_remount=False)

import sys
import subprocess
import importlib
from pathlib import Path
from importlib.machinery import PathFinder

REPO_DIR = Path("/content/drive/MyDrive/Mestrado/PPCOMP_DM")
GITHUB_REPO = "https://github.com/sergiocostaifes/PPCOMP_DM.git"

# Garante o repositório local no Google Drive, clonando apenas se necessário.
if not REPO_DIR.exists():
    REPO_DIR.parent.mkdir(parents=True, exist_ok=True)
    subprocess.run(["git", "clone", GITHUB_REPO, str(REPO_DIR)], check=True)

# Garante prioridade do repositório no sys.path.
repo_str = str(REPO_DIR)
if repo_str in sys.path:
    sys.path.remove(repo_str)
sys.path.insert(0, repo_str)

# Evita uso de módulos antigos em cache no ambiente Colab.
importlib.invalidate_caches()

if PathFinder not in sys.meta_path:
    sys.meta_path.append(PathFinder)

# Garante que src seja tratado como pacote Python.
(REPO_DIR / "src").mkdir(parents=True, exist_ok=True)
init_file = REPO_DIR / "src" / "__init__.py"
if not init_file.exists():
    init_file.write_text("# src package\n", encoding="utf-8")

print("REPO_DIR:", REPO_DIR)

# ─────────────────────────────────────────────────────────────
# BLOCO 2 — Imports do projeto, paths e bibliotecas
# ─────────────────────────────────────────────────────────────

import os
import json
import pandas as pd
import numpy as np
import importlib.util

from importlib import reload

import src.paths as _paths
reload(_paths)

from src.paths import (
    RAW_PATH,
    PROCESSED_PATH,
    FEATURES_PATH,
    MODELS_PATH,
    REPORTS_PATH,
    ensure_dirs,
)

ensure_dirs()

def log(msg: str) -> None:
    """Imprime mensagens padronizadas desta etapa do pipeline."""
    print(f"[01_ingest_validate] {msg}")

print("Paths carregados:")
print("RAW_PATH:", RAW_PATH)
print("PROCESSED_PATH:", PROCESSED_PATH)

# ─────────────────────────────────────────────────────────────
# BLOCO 3 — Arquivo de entrada e leitura diagnóstica do CSV bruto
# ─────────────────────────────────────────────────────────────

CSV_FILE = RAW_PATH / "borg_traces_data.csv"

if not CSV_FILE.exists():
    log(f"Arquivo não encontrado: {CSV_FILE}")
    log(f"Listando RAW_PATH: {RAW_PATH}")
    try:
        log(str(os.listdir(RAW_PATH)))
    except Exception as e:
        log(f"Falha ao listar RAW_PATH: {e}")
    raise FileNotFoundError(f"Arquivo não encontrado: {CSV_FILE}")

log(f"Arquivo encontrado: {CSV_FILE}")

# Primeira leitura: leitura diagnóstica completa do CSV bruto.
df = pd.read_csv(
    CSV_FILE,
    low_memory=False
)

log(f"Leitura OK: {df.shape} (linhas, colunas)")
log(f"Colunas: {list(df.columns)}")

# Remove coluna auxiliar de índice, quando presente no CSV.
df = df.drop(columns=["Unnamed: 0"], errors="ignore")

# ─────────────────────────────────────────────────────────────
# BLOCO 4 — Validação temporal inicial
# ─────────────────────────────────────────────────────────────

if "time" not in df.columns:
    raise KeyError("Coluna 'time' não encontrada no dataset.")

# Converte time para numérico e transforma valores sentinela em NaN.
df["time"] = pd.to_numeric(df["time"], errors="coerce")

sentinel_cutoff = 1e16

n_before = int(df["time"].isna().sum())
df.loc[df["time"] > sentinel_cutoff, "time"] = np.nan
n_after = int(df["time"].isna().sum())

log(f"time NaNs antes: {n_before} | depois de sentinela: {n_after}")
log(f"time min: {df['time'].min()}")
log(f"time median: {df['time'].median()}")
log(f"time max: {df['time'].max()}")

if df["time"].min() == 0:
    log(
        "Observação: time==0 será tratado no 02_clean_normalize "
        "(remoção/ajuste conforme regra do projeto)."
    )

# ─────────────────────────────────────────────────────────────
# BLOCO 5 — Consistência semântica entre event e failed
# ─────────────────────────────────────────────────────────────

# Verifica se a codificação binária de falha é coerente com o evento textual FAIL.
if "event" in df.columns and "failed" in df.columns:
    concord = (df["event"].eq("FAIL") == df["failed"].eq(1)).mean()
    log(f"Concordância event=='FAIL' vs failed==1: {concord:.4f}")
else:
    log("Colunas event/failed não encontradas; pulando checagem de concordância.")

# ─────────────────────────────────────────────────────────────
# BLOCO 6 — Summary diagnóstico do CSV bruto
# ─────────────────────────────────────────────────────────────

# Este summary registra informações essenciais da ingestão.
# A AED detalhada do dataset bruto é responsabilidade do NB00.
summary = {
    "raw_rows": int(df.shape[0]),
    "raw_cols": int(df.shape[1]),
    "raw_time_nan": int(df["time"].isna().sum()),
    "raw_failed_1_count": int((df["failed"] == 1).sum()) if "failed" in df.columns else None,
    "raw_event_top": df["event"].value_counts().head(10).to_dict() if "event" in df.columns else None,
}

# Em abril, este dicionário era exibido como saída de uma célula própria.
# Em célula única, esta exibição explícita preserva a visibilidade diagnóstica.
display(summary)

# ─────────────────────────────────────────────────────────────
# BLOCO 7 — Seleção de colunas e preparação do Parquet validado
# ─────────────────────────────────────────────────────────────

# Confirma disponibilidade do pyarrow para persistência em Parquet.
if importlib.util.find_spec("pyarrow") is None:
    subprocess.run(
        [sys.executable, "-m", "pip", "-q", "install", "pyarrow"],
        check=True
    )

# Seleção estendida de colunas mantidas para o pipeline.
# Esta seleção preserva a versão histórica de abril.
KEEP_COLS = [  # EXTENDED
    "time", "failed", "event", "machine_id", "collection_id", "instance_index",
    "priority", "scheduling_class", "start_time", "end_time",
    "resource_request", "average_usage", "maximum_usage", "random_sample_usage",
    "assigned_memory", "page_cache_memory",
    "cycles_per_instruction", "memory_accesses_per_instruction",
    "sample_rate", "cluster", "scheduler"
]

# Tipagem histórica da versão de abril.
# Atenção: aqui não se deve alterar para manter coerência com a baseline do artigo.
CAT_COLS = [
    c for c in ["event", "cluster", "scheduler", "scheduling_class", "priority"]
    if c in KEEP_COLS
]

FLOAT32_COLS = [
    c for c in [
        "average_usage", "maximum_usage", "random_sample_usage",
        "assigned_memory", "page_cache_memory",
        "cycles_per_instruction", "memory_accesses_per_instruction",
        "sample_rate",
    ]
    if c in KEEP_COLS
]

INT_COLS = [
    c for c in ["machine_id", "collection_id", "instance_index", "start_time", "end_time"]
    if c in KEEP_COLS
]

# Segunda leitura: releitura do CSV apenas com as colunas necessárias ao pipeline.
df = pd.read_csv(
    CSV_FILE,
    usecols=KEEP_COLS,
    low_memory=False
)

# Tratamento de time e valores sentinela.
df["time"] = pd.to_numeric(df["time"], errors="coerce")
sentinel_cutoff = 1e16
df.loc[df["time"] > sentinel_cutoff, "time"] = np.nan

# Conversão da variável binária de falha.
df["failed"] = (
    pd.to_numeric(df["failed"], errors="coerce")
    .fillna(0)
    .astype("int8")
)

# Conversão de colunas categóricas conforme a versão histórica de abril.
for c in CAT_COLS:
    df[c] = df[c].astype("category")

# Conversão de inteiros nullable.
for c in INT_COLS:
    df[c] = pd.to_numeric(df[c], errors="coerce").astype("Int64")

# Conversão de numéricos em float32 para reduzir uso de memória.
for c in FLOAT32_COLS:
    df[c] = pd.to_numeric(df[c], errors="coerce").astype("float32")

# Remove apenas registros sem time válido após o tratamento de sentinela.
# Registros com time == 0 são preservados.
n_before = len(df)
df = df.dropna(subset=["time"]).copy()
df["time"] = df["time"].astype("int64")

print(f"Drop time NaN: {n_before} -> {len(df)}")

# ─────────────────────────────────────────────────────────────
# BLOCO 8 — Persistência do dataset validado
# ─────────────────────────────────────────────────────────────

VALIDATED_FILE = PROCESSED_PATH / "trace_raw_validated.parquet"

df.to_parquet(
    VALIDATED_FILE,
    engine="pyarrow",
    compression="snappy",
    index=False
)

log(f"Salvo: {VALIDATED_FILE}")
log(f"Tamanho (MB): {round(VALIDATED_FILE.stat().st_size / 1024 / 1024, 2)}")

# ─────────────────────────────────────────────────────────────
# BLOCO 9 — Atualização e persistência do summary
# ─────────────────────────────────────────────────────────────

summary["parquet_rows"] = int(len(df))
summary["parquet_cols"] = int(df.shape[1])
summary["kept_cols"] = list(df.columns)
summary["validated_file"] = str(VALIDATED_FILE)

summary_file = REPORTS_PATH / "01_ingest_validate_summary.json"

summary_file.write_text(
    json.dumps(summary, indent=2, ensure_ascii=False),
    encoding="utf-8"
)

log(f"Resumo salvo: {summary_file}")

Mounted at /content/drive
REPO_DIR: /content/drive/MyDrive/Mestrado/PPCOMP_DM
Paths carregados:
RAW_PATH: /content/drive/MyDrive/Mestrado/02-datasets/01-raw
PROCESSED_PATH: /content/drive/MyDrive/Mestrado/02-datasets/02-processed
[01_ingest_validate] Arquivo encontrado: /content/drive/MyDrive/Mestrado/02-datasets/01-raw/borg_traces_data.csv
[01_ingest_validate] Leitura OK: (405894, 34) (linhas, colunas)
[01_ingest_validate] Colunas: ['Unnamed: 0', 'time', 'instance_events_type', 'collection_id', 'scheduling_class', 'collection_type', 'priority', 'alloc_collection_id', 'instance_index', 'machine_id', 'resource_request', 'constraint', 'collections_events_type', 'user', 'collection_name', 'collection_logical_name', 'start_after_collection_ids', 'vertical_scaling', 'scheduler', 'start_time', 'end_time', 'average_usage', 'maximum_usage', 'random_sample_usage', 'assigned_memory', 'page_cache_memory', 'cycles_per_instruction', 'memory_accesses_per_instruction', 'sample_rate', 'cpu_usage_distr

{'raw_rows': 405894,
 'raw_cols': 33,
 'raw_time_nan': 3,
 'raw_failed_1_count': 92678,
 'raw_event_top': {'FINISH': 92867,
  'FAIL': 92678,
  'ENABLE': 75907,
  'SCHEDULE': 69104,
  'LOST': 59515,
  'EVICT': 14756,
  'KILL': 951,
  'UPDATE_PENDING': 111,
  'QUEUE': 4,
  'UPDATE_RUNNING': 1}}

Drop time NaN: 405894 -> 405891
[01_ingest_validate] Salvo: /content/drive/MyDrive/Mestrado/02-datasets/02-processed/trace_raw_validated.parquet
[01_ingest_validate] Tamanho (MB): 12.3
[01_ingest_validate] Resumo salvo: /content/drive/MyDrive/Mestrado/04-reports/01_ingest_validate_summary.json


# Conclusão da Ingestão e Validação dos Dados

A etapa de ingestão e validação foi concluída com sucesso.

O NB01 realizou a leitura do arquivo bruto `borg_traces_data.csv`, removeu a coluna auxiliar `Unnamed: 0`, validou a estrutura inicial do dataset, tratou valores temporais inválidos associados a sentinelas e persistiu uma versão validada e otimizada dos dados para uso nas etapas seguintes do pipeline.

Nesta etapa não foram realizadas transformações analíticas, agregação temporal, detecção de episódios, engenharia de atributos ou modelagem. O foco permaneceu na confiabilidade estrutural da base de entrada.

## 8. Resultados da Execução

A execução atual apresentou os seguintes resultados:

| Item | Valor |
|---|---:|
| Registros lidos no CSV bruto | 405.894 |
| Colunas após remoção de `Unnamed: 0` | 33 |
| Registros com `time` inválido após tratamento de sentinela | 3 |
| Registros persistidos em `trace_raw_validated.parquet` | 405.891 |
| Colunas persistidas no Parquet validado | 21 |

A variável temporal apresentou os seguintes valores de referência:

| Estatística | Valor |
|---|---:|
| Mínimo de `time` | 0 |
| Mediana de `time` | 1.084.786.677.422 |
| Máximo de `time` | 2.678.923.967.375 |

A presença de registros com `time == 0` foi apenas registrada nesta etapa. A decisão metodológica sobre preservação, interpretação e tratamento do bloco inicial da série temporal é tratada no NB02 e retomada nos notebooks posteriores de agregação e análise exploratória.

## 9. Consistência Semântica das Falhas

Foi verificada concordância total entre:

- `event == "FAIL"`;
- `failed == 1`.

A concordância observada foi igual a **1,0000**.

Esse resultado indica que a variável binária `failed` está semanticamente coerente com a categoria textual `FAIL` da variável `event`. Portanto, seu uso nas etapas posteriores de agregação temporal, cálculo de taxa de falha e detecção de criticidade é consistente com a semântica original dos eventos.

Os eventos mais frequentes observados no dataset bruto foram:

| Evento | Registros |
|---|---:|
| `FINISH` | 92.867 |
| `FAIL` | 92.678 |
| `ENABLE` | 75.907 |
| `SCHEDULE` | 69.104 |
| `LOST` | 59.515 |
| `EVICT` | 14.756 |
| `KILL` | 951 |
| `UPDATE_PENDING` | 111 |
| `QUEUE` | 4 |
| `UPDATE_RUNNING` | 1 |

A distribuição confirma a diversidade operacional do dataset e a presença de volume relevante de eventos de falha.

## 10. Colunas Mantidas no Dataset Validado

O arquivo `trace_raw_validated.parquet` foi persistido com 21 colunas selecionadas para sustentar as próximas etapas do pipeline.

As colunas mantidas foram:

| Grupo | Colunas |
|---|---|
| Temporal | `time`, `start_time`, `end_time` |
| Evento e falha | `event`, `failed` |
| Identificação operacional | `machine_id`, `collection_id`, `instance_index`, `cluster` |
| Priorização e escalonamento | `priority`, `scheduling_class`, `scheduler` |
| Recursos solicitados e uso | `resource_request`, `average_usage`, `maximum_usage`, `random_sample_usage` |
| Memória e amostragem | `assigned_memory`, `page_cache_memory`, `sample_rate` |
| Métricas de instrução | `cycles_per_instruction`, `memory_accesses_per_instruction` |

A seleção preserva os campos necessários para normalização temporal, agregação por janelas, cálculo de métricas operacionais e construção posterior de atributos.

## 11. Artefatos Gerados

Nesta execução, foram gerados os seguintes artefatos:

| Artefato | Finalidade |
|---|---|
| `trace_raw_validated.parquet` | Base validada e otimizada para as etapas posteriores do pipeline |
| `01_ingest_validate_summary.json` | Resumo rastreável da ingestão e validação |

O arquivo `trace_raw_validated.parquet` contém a base validada, com tipos ajustados para reduzir uso de memória e preservar os campos relevantes para os notebooks seguintes.

O arquivo `01_ingest_validate_summary.json` registra os principais indicadores da execução, incluindo número de registros lidos, registros removidos por tempo inválido, colunas mantidas e estatísticas básicas da ingestão.

## 12. Encaminhamento para o Pipeline

Com os dados brutos validados, o pipeline segue para o NB02, responsável pela limpeza e normalização temporal.

O NB02 deverá partir do arquivo `trace_raw_validated.parquet`, preservar a rastreabilidade da variável `time` e documentar explicitamente a decisão sobre o início da série temporal, especialmente em relação aos registros com `time == 0`.

A partir da base validada no NB01, os notebooks seguintes poderão:

1. normalizar a estrutura temporal;
2. construir a série agregada em janelas de 5 minutos;
3. calcular a taxa de falha por janela;
4. detectar janelas críticas e episódios;
5. gerar atributos temporais;
6. rotular estados operacionais;
7. formular a base supervisionada de antecipação.

## 13. Síntese Final

O NB01 cumpre adequadamente seu papel como etapa de ingestão e validação estrutural do pipeline.

A execução preservou praticamente a totalidade dos registros do dataset bruto, removendo apenas os registros com valor temporal inválido após tratamento de sentinela. A concordância perfeita entre `event == "FAIL"` e `failed == 1` reforça a consistência semântica da variável de falha utilizada nas etapas posteriores.

O notebook deve, portanto, ser considerado fechado como etapa de preparação inicial dos dados, sem necessidade de reexecução para fins de revisão editorial. Os ajustes recomendados devem permanecer restritos à organização do markdown e à consolidação visual do código, preservando os artefatos já utilizados pelo restante do pipeline.